File: 874. Walking Robot Simulation.py
class Solution:
    def robotSim(self, commands: List[int], obstacles: List[List[int]]) -> int:
        # Convert the list of obstacles to a set for O(1) lookup time
        obstacle_set = set(map(tuple, obstacles))
        
        # Directions: North, East, South, West (in terms of x, y changes)
        directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
        
        # Initial position and direction (facing North)
        x, y = 0, 0
        direction_index = 0
        
        max_distance_squared = 0
        
        for command in commands:
            if command == -2:
                # Turn left (counterclockwise)
                direction_index = (direction_index - 1) % 4
            elif command == -1:
                # Turn right (clockwise)
                direction_index = (direction_index + 1) % 4
            else:
                # Move forward 'command' units
                for _ in range(command):
                    # Calculate next position
                    next_x = x + directions[direction_index][0]
                    next_y = y + directions[direction_index][1]
                    
                    # Check if the next position is an obstacle
                    if (next_x, next_y) in obstacle_set:
                        # If there's an obstacle, stop moving in this direction
                        break
                    
                    # Update position
                    x, y = next_x, next_y
                    
                    # Calculate the square of the distance from the origin
                    max_distance_squared = max(max_distance_squared, x * x + y * y)
        
        return max_distance_squared

File: 874. Walking Robot Simulation.cpp
#include <memory_resource>
pmr::unsynchronized_pool_resource pool;
class Solution {
public:
    int robotSim(vector<int>& commands, vector<vector<int>>& obstacles) {
        const unsigned lb=30000;
        pmr::unordered_set<unsigned> obSet(&pool);
        obSet.reserve(obstacles.size());
        for(auto& ob: obstacles){
            const unsigned x=ob[0]+lb, y=ob[1]+lb;
            obSet.insert((x<<16)+y);
        }

        const int dir[4][2]={{0, 1}, {-1, 0}, {0, -1}, {1, 0}};
        int x=0, y=0, dx=0, dy=1, face=0, maxD2=0;
        for(int c: commands){
            switch(c){
                case -2: face=(face+1)&3; dx=dir[face][0]; dy=dir[face][1]; break;
                case -1: face=(face+3)&3; dx=dir[face][0]; dy=dir[face][1]; break;
                default:
                //    cout<<"\nc="<<c<<" face="<<face<<":";
                    for(int i=0; i<c; i++){
                        x+=dx, y+=dy;
                        if (obSet.count(((x+lb)<<16)+y+lb)) {
                            x-=dx;  // previous move
                            y-=dy;
                            break;
                        }
                    //    cout<<"("<<x<<","<<y<<"),";
                        maxD2=max(maxD2, x*x+y*y);
                    }
            }
        }
        return maxD2;
    }
};
